### PR TITLE
kompile: add check for token sort productions

### LINKED ISSUE #2119: [kompile] should give more readable error message when rejecting token sorts with constructors
Right now we get a hard to read error message from the haskell backend, and a crash from the llvm backend:
```
module TEST
  syntax Foo ::= foo()
  syntax Foo ::= bam()
  syntax Foo ::= fu()
  syntax Foo ::= "bar" [token]
               | r"ba." [token, prec(3)]

  rule foo() => .K
  rule bam() => .K
  rule fu() => .K
  rule bar => .K
  rule baz => .K
endmodule

$ kompile test.k
[Error] Internal: Uncaught exception thrown of type MatchError.
Please rerun your program with the --debug flag to generate a stack trace, and
file a bug report at https://github.com/kframework/k/issues (MatchError:
(LiteralC(bar),Lblbam'LParRParUnds'TEST'Unds'Foo{}) (of class scala.Tuple2))

$ kompile test.k --backend haskell
Error:
  module 'TEST':
  symbol 'Lblbam'LParRParUnds'TEST'Unds'Foo' declaration (definition.kore 113:10):
    Cannot define constructor Lblbam'LParRParUnds'TEST'Unds'Foo for sort SortFoo{}: sort has domain values.
[Error] Critical: Haskell backend reported errors validating compiled
definition.
Examine output to see errors.
```

### PR DESCRIPTION
**Pull request by SchmErik**
Token sorts should only contain productions that have `token` or
`function` labels. Any other label (including subsorts) are illegal.

Fixes: #2119

### CODE REVIEW

File: kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import org.kframework.definition.Production;
+import org.kframework.definition.Sentence;
+import org.kframework.utils.errorsystem.KEMException;
+
+import java.util.Set;
+
+public class CheckTokens {
+    private final Set<KEMException> errors;
+    private final Module m;
+    private static final ImmutableSet<String> ignoredSortNames = ImmutableSet.of("Bag", "IOFile", "IOInt", "IOString", "K", "KBott", "KItem", "KLabel", "KList", "Sort");
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java]: Why `Sort`? Where is that coming from?
Not sure about some others also. `KItem` for example.
I just noticed that in the original issue, I complained about the LLVM and Haskell backend. But Java is a bit different. Probably we don't need to run this check there.
Could you check all these, please?

File: kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    public CheckTokens(Set<KEMException> errors, Module m) {
+        this.errors = errors;
+        this.m = m;
+    }
+
+    // This check ensures that sorts containing token declarations only contain syntax declarations that are also tokens (or macros).
+    public void check(Sentence sentence) {
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java:23]: `private void check(Production p) {`
You can match directly on Production and do away with this check.

File: kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                || ignoredSortNames.contains(p.sort().name())) {
+            return;
+        }
+
+        if (!m.tokenProductionsFor().contains(p.sort()) // We only care about sorts that have been declared as tokens.
+                || p.klabel().isDefined() && m.macroKLabels().contains(p.klabel().get())) {
+            return;
+        }
+
+        errors.add(KEMException.compilerError("Sort " + p.sort().name() + " was declared as a token. Productions of this sort can only contain [function] or [token] labels.", p));
```
Comment (radumereuta) [inline kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java]: Line too long.

**Review by radumereuta — CHANGES_REQUESTED** [review]

File: kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java (hunk 2 above)
Comment (SchmErik) [reply kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java:23]: ah, thanks that would make things more clear

File: kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java (hunk 1 above)
Comment (SchmErik) [reply kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java]: I came up with these by running kompile on the tokenCheck.k file and looking at the errors that were generated without ignoring special sort names and `Sort` was one of them

Comment (radumereuta) [reply kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java]: You should check all of them. They might indicate something else is wrong.

Comment (SchmErik) [reply kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java]: oh strange... Now I'm getting that I only need to ignore "KBott" and "KLabel"

Comment (SchmErik) [reply kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java]: I think this was written when I was still working on solidifying the logic. I removed all but KBott and KLabel and I'm rerunning the testsuite locally. I'll push another commit addressing this and the other feedback you have. Thanks!

**Comment by SchmErik** [comment]
this PR got a little hectic with a bunch of commits that ended up getting reverted or were typo fixes. Now that the CI is passing, I'll open yet another PR with cleaner commits.

**Pull request closed without merging**